TypeCobol/Compiler/CodeModel/SymbolTable.cs

```diff
@@ -57,40 +31,63 @@ public SymbolTable(SymbolTable enclosing, Scope current)
                 throw new InvalidOperationException("Only Table of INTRINSIC symbols don't have any enclosing scope.");
         }
 
-        /// <summary>
-        /// GetFromTableAndEnclosing 2 
-        /// </summary>
-        /// <typeparam name="T"></typeparam>
-        /// <param name="head"></param>
-        /// <param name="getTableFunction"></param>
-        /// <param name="maxScope"></param>
-        /// <returns></returns>
-        [NotNull]
-        private List<T> GetFromTableAndEnclosing<T>(string head,
-            Func<SymbolTable, IDictionary<string, List<T>>> getTableFunction, Scope maxScope = Scope.Intrinsic) where T : Node
+        private delegate void MatchDelegate<TNode>(List<TNode> found,
+            in TNode head,
+            in QualifiedName name,
+            int nameIndex,
+            in TNode current)
+            where TNode : Node;
+
+        private List<T> GetFromTableAndEnclosing<T>(QualifiedName name,
+            Func<SymbolTable, IDictionary<string, List<T>>> getEntries,
+            MatchDelegate<T> match,
+            Scope maxScope,
+            Scope? requiredScope = null)
+            where T : Node
         {
-            System.Diagnostics.Debug.Assert(head != null);
+            System.Diagnostics.Debug.Assert(requiredScope == null || requiredScope <= maxScope);
+
             var result = new List<T>();
-            this.GetFromTableAndEnclosing2(head, getTableFunction, result, maxScope);
-            
+            bool skipParentGlobal = false;
+
+            // Iterate over scopes from current up to maxScope
+            var currentSymbolTable = this;
+            while (currentSymbolTable != null && currentSymbolTable.CurrentScope >= maxScope)
+            {
+                // Retrieve candidates from current SymbolTable
+                if (!getEntries(currentSymbolTable).TryGetValue(name.Head, out var candidates)) candidates = new List<T>();
```

Comment:
the tests cannot be combined because either we have entries with correct name or not, we have to update the skipGlobal boolean.
I suggest this code a bit more efficient :
```
// Retrieve candidates from current SymbolTable
getEntries(currentSymbolTable).TryGetValue(name.Head, out var candidates);

// We don't have to perform matching if we already have found a result (except if current scope is required)
if (!skipParentGlobal || currentSymbolTable.CurrentScope == requiredScope)
{
    // Proceed to matching
    if (candidates != null)
    {
        foreach (var candidate in candidates)
        {
            match(result, candidate, name, name.Count - 1, candidate);
        }
    }

    // Skip Global SymbolTable from parent if we already have found a match
    skipParentGlobal = currentSymbolTable.CurrentScope == Scope.Global && result.Count > 0;
}
```